### DISCUSSION TITLE
benchmark for interprocessor comunication

### DISCUSSION THREAD

**Discussion by Ankita-020696** [OpenAMP/open-amp#318] [Q&A]
Hi,
I want to use interprocessor communication between rasberrypi and i.mx boards. How do i start with it? Also is there any open source benchmark application for inter processor communication?

Thanks

**Comment by arnopo** [comment]
Hello 
If i well understood you want to have a communication between two boards, right?
In this case what would be the link for the IPC?

**Reply by Ankita-020696** [reply]
> Hello If i well understood you want to have a communication between two boards, right? In this case what would be the link for the IPC?

Yes, how shall I start with communication between board?
Like how to install openamp library on RPI/i.mx?
How to use it for communication?

Thanks

**Reply by arnopo** [reply]
> Yes, how shall I start with communication between board? Like how to install openamp library on RPI/i.mx? How to use it for communication?

The OpenAMP lib implements the RPMsg protocol based on a Virtio transport layer that need shared memory and intercore signaling (mailbox).
If you plan to exchange date using a low rate link such as an UART you will need to create a backend to replace the virtio One in OpenAMP and/or in Linux Kernel.

The alternative is to directly implement your own protocol to communicate direcly through an UART link, probably simpler to implement.

The only example i know of board communication based on RPMsg is the demo I did 2 years ago.  This demo is based on a communication between a stm32MP1 board running a Linux with a stm32L4 IOT board running a Zephyr.
- the linux Kerbnel has been patched to integrate a RPMsg over UART backend
- the Zephyr has been patched to implement RPMsg over UART ( no use of the OpenAMP lib). 

Details is avalable here: https://github.com/arnopo?tab=projects

**Comment by FirstLoveLife** [comment]
Hi Arnopo,

> The only example i know of board communication based on RPMsg is the demo I did 2 years ago. This demo is based on a communication between a stm32MP1 board running a Linux with a stm32L4 IOT board running a Zephyr.
> 
> * the linux Kerbnel has been patched to integrate a RPMsg over UART backend
> * the Zephyr has been patched to implement RPMsg over UART ( no use of the OpenAMP lib).

How does your two baords notify each other? mailbox? platform msi? software interrupt? or just poll address in shared memory?

> Details is avalable here: https://github.com/arnopo?tab=projects

The projects are not available now, could you help re-post them? Thanks.

**Comment by arnopo** [comment]
> Hi Arnopo,
> 
> > The only example i know of board communication based on RPMsg is the demo I did 2 years ago. This demo is based on a communication between a stm32MP1 board running a Linux with a stm32L4 IOT board running a Zephyr.
> > 
> > * the linux Kerbnel has been patched to integrate a RPMsg over UART backend
> > * the Zephyr has been patched to implement RPMsg over UART ( no use of the OpenAMP lib).
> 
> How does your two baords notify each other? mailbox? platform msi? software interrupt? or just poll address in shared memory?

If you're talking about communication between two processors on two separate boards, you probably can't use shared memory. 
We use existing UART interrupts  for the notifications. 

> 
> > Details is avalable here: https://github.com/arnopo?tab=projects
> 
> The projects are not available now, could you help re-post them? Thanks.

The project description is in the "classic" tab: https://github.com/arnopo?tab=projects&type=classic

**Comment by FirstLoveLife** [comment]
> > Hi Arnopo,
> > > The only example i know of board communication based on RPMsg is the demo I did 2 years ago. This demo is based on a communication between a stm32MP1 board running a Linux with a stm32L4 IOT board running a Zephyr.
> > > 
> > > * the linux Kerbnel has been patched to integrate a RPMsg over UART backend
> > > * the Zephyr has been patched to implement RPMsg over UART ( no use of the OpenAMP lib).
> > 
> > 
> > How does your two baords notify each other? mailbox? platform msi? software interrupt? or just poll address in shared memory?
> 
> If you're talking about communication between two processors on two separate boards, you probably can't use shared memory. We use existing UART interrupts for the notifications.

I am referring to the communication between two processors on two different boards. Are you suggesting that sending UART data can trigger the UART interrupt on the remote board? I am not quite grasping the concept. Could you please provide more information on the specifics? Thank you.

> > > Details is avalable here: https://github.com/arnopo?tab=projects
> > 
> > 
> > The projects are not available now, could you help re-post them? Thanks.
> 
> The project description is in the "classic" tab: https://github.com/arnopo?tab=projects&type=classic

Thanks!
